File: dsa/topics/19_bit/java/L371_SumOfTwoIntegers.java
public class L371_SumOfTwoIntegers {

    public static void main(String[] args) {
        L371_SumOfTwoIntegers obj = new L371_SumOfTwoIntegers();

        System.out.println(obj.getSum(1, 2));
        System.out.println(obj.getDiff(1, 2));
    }

    // Problem: https://leetcode.com/problems/sum-of-two-integers/
    // Idea: In binary addition, for 1,0 0,1 0,0 1,1 the sum will be 1,1,0,0. This we can get using xor.
    // But 1,1 case generates a carry which needs to be added to next bit sum.
    // When both are 1, the carry need to be 1. Hence use & to generate carry.
    // But the carry need to be added to next bit. Hence right shift it by one bit.
    // Repeate is until carry becomes 0.
    public int getSum(int a, int b) {
        while (b != 0) {
            int carry = a & b;
            a = a ^ b;          // sum
            b = carry << 1;     // carry after shifting
        }

        return a;
    }

    public int getDiff(int a, int b) {
        while (b != 0) {
            int carry = (~a) & b;
            a = a ^ b;
            b = carry << 1;
        }
        return a;
    }
}

File: dsa/topics/19_bit/python/l371_sum_of_two_integers.py
# Problem: https://leetcode.com/problems/sum-of-two-integers/
# Idea: In binary addition, for 1,0 0,1 0,0 1,1 the sum will be 1,1,0,0. This we can get using xor.
# But 1,1 case generates a carry which needs to be added to next bit sum.
# When both are 1, the carry need to be 1. Hence use & to generate carry.
# But the carry need to be added to next bit. Hence right shift it by one bit.
# Repeate is until carry becomes 0.
def get_sum(a, b):
    MASK = 0xffffffff
    INT_MAX = 0x7fffffff

    while b != 0:
        carry = (a & b) & MASK
        a = (a ^ b) & MASK     # sum
        b = (carry << 1) & MASK  # carry after shifting

    return a if a <= INT_MAX else ~(a ^ MASK)


def get_diff(a, b):
    MASK = 0xffffffff
    INT_MAX = 0x7fffffff

    while b != 0:
        carry = ((~a) & b) & MASK
        a = (a ^ b) & MASK
        b = (carry << 1) & MASK

    return a if a <= INT_MAX else ~(a ^ MASK)


# ------------------ Test (Same as Java main) ------------------

print(get_sum(1, 2))   # Expected: 3
print(get_diff(1, 2))  # Expected: -1
